0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+  print s[19], "\n" // /unicodeChar:0 => "�"
+                    // /unicodeChar:1 => "😎"
+}
+```
+
+A related but largely orthogonal issue has also been addressed.
+Dafny source code files must be encoded in UTF-8,
+but in previous versions of Dafny,
+string and character literals could only contain printable and white-space ASCII characters,
+due to a limitation of the Coco/R parser generator the toolchain uses.
+This has been fixed, and both standard form and verbatim form string literals now allow any Unicode characters.
+A second form of escape sequence accepting a hexadecimal number with up to six digits, `\u{XXXXXX}`, 
+is now provided to support characters outside of the Basic Multilingual Plane
+using their direct Unicode code points instead of using surrogates.
+These changes are fully backwards-compatible and not controlled by the `unicodeChar` flag.
+
+```dafny
+// Several different ways to express the same string literal
+var s1 := "Unicode is just so \ud83d\ude0e";
+var s2 := "Unicode is just so \u{1F60E}";
+var s3 := "Unicode is just so 😎";
+var s4 := @"Unicode is just so 😎";  // Escape sequences are not supported in verbatim strings
+```
+
+The exact representation of strings at runtime, including the particular encoding,
+is an implementation detail of a particular backend, and will often be optimized for the idioms and support
+of the target environment. Enabling Unicode characters will change the target language types used to
+represent characters and strings, and hence may cause compilation errors when using additional external
+target language code.
+
+Note also that although the Unicode scalar value concept is more general than UTF-16 code units,
+it still does not always correspond to what humans will perceive as single atomic characters when rendered.
+For example, the string `"e\u0301"` contains two Unicode scalar values, but renders as the single character `é`.
+See the concept of grapheme clusters [here](https://unicode.org/reports/tr29/) for more details.
+The proposed change to the `char` type is only intended to allow the Dafny language to safely abstract
+away from encodings, especially to support verifiably-correct code that must compile to multiple target languages.
+Providing more of the concepts defined by the Unicode standard is left out of scope for this proposal,
+under the assumption that it will enable such implementations via Dafny source code in sharable libraries instead.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Verification
+
+The change to the definition of `char` in the `DafnyPrelude.bpl` file is minimal.
+The only change is to replace expressions such as `0 <= n && n < 65536` with a separate predicate defined as follows:
+
+```boogie
+function char#IsUnicodeScalarValue(n: int): bool {
+  (0                  <= n && n <= 55295   /* 0xD7FF */) || 
+  (57344 /* 0xE000 */ <= n && n <= 1114111 /* 0x10FFFF */ )
```

Comment:
Use half-open intervals.

Reply:
The only reason I used closed intervals throughout is because that's what the Unicode standard itself does. It's always irked me, but as a result the constant `0xDFFF` tends to be much more well-known than `0xE000`.

I will absolutely use half-open intervals here as it will probably verify more efficiently, and I'm happy to change the proposal accordingly if it would help readability.

Reply:
Ended up going with half-open intervals for consistency.